File: src/tools/agent.ts
import type { ToolDef, ToolResult, ToolExecutionContext } from './index.js';
import type { AgentConfig } from '../agent.js';
import type { MessageParam, ContentBlock } from '../llm/anthropic.js';
import { Agent } from '../agent.js';
import { ConsoleDisplay } from '../utils/display.js';

export const agentTool: ToolDef = {
  name: 'Agent',
  description: 'Delegate a sub-task to an independent agent. Creates a new agent session that runs in parallel. The sub-agent has access to all tools (except agent) and returns a concise summary.',
  requires: ['agentRegistry'],
  input_schema: {
    type: 'object' as const,
    properties: {
      task: {
        type: 'string' as const,
        description: 'The specific task to delegate to the sub-agent'
      }
    },
    required: ['task']
  },
  execute: async (args: Record<string, unknown>, context?: ToolExecutionContext): Promise<ToolResult> => {
    const task = args.task as string;
    const registry = context?.registry;
    const config = context?.config;
    const parentId = context?.currentAgentId || '1';

    if (!registry) {
      return { output: 'Error: AgentRegistry not available' };
    }

    if (!config) {
      return { output: 'Error: Agent config not available' };
    }

    const subId = registry.allocateSubId();

    const subConfig: AgentConfig = {
      ...config,
      display: new ConsoleDisplay(),
      excludeTools: ['Agent'],
      agentRegistry: registry,
      currentAgentId: subId,
    };

    const subAgent = new Agent(subConfig);

    context?.signal?.addEventListener('abort', () => {
      subAgent.abort();
    });

    registry.register({
      id: subId,
      type: 'sub',
      agent: subAgent,
      status: 'running',
      task,
      parentId,
    });

    // Notify via parent agent's store
    const parentSession = registry.get(parentId);
    if (parentSession) {
      const taskPreview = task.length > 40 ? task.slice(0, 40) + '...' : task;
      parentSession.agent.getStore().addStatus({
        role: 'status',
        content: `[Agent #${subId} started: ${taskPreview}]`,
        timestamp: new Date(),
      });
    }

    try {
      await subAgent.run(task);
      const turns = subAgent.getStore().getTurns();
      const finalResponse = extractFinalResponse(turns);
      const summary = generateSummary(turns);
      registry.updateStatus(subId, 'completed');
      registry.updateSummary(subId, summary);
      registry.remove(subId);

      const output = finalResponse || `Agent #${subId} completed: ${summary}`;
      return { output };
    } catch (error) {
      const errorMsg = error instanceof Error ? error.message : String(error);
      registry.updateStatus(subId, 'error');
      registry.updateSummary(subId, `Error: ${errorMsg}`);
      registry.remove(subId);

      return { output: `Agent #${subId} failed: ${errorMsg}` };
    }
  }
};

function extractFinalResponse(turns: MessageParam[]): string | null {
  for (let i = turns.length - 1; i >= 0; i--) {
    const turn = turns[i];
    if (turn.role === 'assistant' && Array.isArray(turn.content)) {
      for (const block of turn.content as ContentBlock[]) {
        if (block.type === 'text' && block.text.trim()) {
          return block.text.trim();
        }
      }
    }
  }
  return null;
}

function generateSummary(turns: MessageParam[]): string {
  let toolCallCount = 0;

  for (const turn of turns) {
    if (turn.role === 'assistant' && Array.isArray(turn.content)) {
      for (const block of turn.content as ContentBlock[]) {
        if (block.type === 'tool_use') toolCallCount++;
      }
    }
  }

  return toolCallCount > 0 ? `${toolCallCount} operations` : 'Task completed';
}

File: internal/tools/agent.go
package tools

import (
	"context"
	"fmt"
	"strings"

	"minicode/internal/config"
	"minicode/internal/domain"
)

// SubAgentTool delegates work to a sub-agent with optional tier-based model override.
type SubAgentTool struct {
	parentConfig domain.AgentConfig
}

// NewSubAgentTool creates a SubAgentTool.
func NewSubAgentTool(parentConfig domain.AgentConfig) *SubAgentTool {
	return &SubAgentTool{parentConfig: parentConfig}
}

func (t *SubAgentTool) Name() string { return "SubAgent" }
func (t *SubAgentTool) Description() string {
	return "Delegate a sub-task to an independent agent. Creates a new agent session. The sub-agent has access to all tools (except SubAgent) and returns a concise summary."
}
func (t *SubAgentTool) ReadOnly() bool    { return false }
func (t *SubAgentTool) Interactive() bool { return false }
func (t *SubAgentTool) Requires() []ToolRequirement { return nil }

func (t *SubAgentTool) InputSchema() map[string]any {
	return map[string]any{
		"type": "object",
		"properties": map[string]any{
			"task": map[string]any{"type": "string", "description": "The specific task to delegate to the sub-agent"},
			"tier": map[string]any{"type": "string", "description": "Optional: run the sub-agent with the model mapped to this tier (\"1\", \"2\", or \"3\")."},
		},
		"required": []string{"task"},
	}
}

func (t *SubAgentTool) Execute(ctx context.Context, args map[string]any, tc ToolContext) (domain.ToolResult, error) {
	task, _ := args["task"].(string)
	tier, _ := args["tier"].(string)
	if task == "" {
		tc.Log("SubAgent: task is required")
		return domain.ToolResult{Output: "Error: task is required"}, nil
	}

	if tc.AgentFactory == nil {
		tc.LogErr("SubAgent: no agent factory available")
		return domain.ToolResult{Output: "Error: No agent factory available"}, nil
	}

	tc.Log("SubAgent: launching sub-agent", "task_length", len(task), "tier", tier)

	subCfg := t.parentConfig

	// Override model from tier mapping
	if tier != "" {
		cfg, err := config.Load()
		if err == nil && cfg.Tiers != nil {
			if modelSpec, ok := cfg.Tiers[tier]; ok {
				resolved, err := config.Resolve(modelSpec)
				if err == nil && resolved.Model.APIKey != "" {
					subCfg.Model = resolved.Model.Model
					subCfg.APIKey = resolved.Model.APIKey
					subCfg.BaseURL = resolved.Model.BaseURL
					subCfg.ContextLength = resolved.Model.ContextLength
					tc.Log("SubAgent: using tier model", "tier", tier, "model", subCfg.Model)
				}
			}
		}
	}

	subAgent := tc.AgentFactory.Create(subCfg)
	err := tc.AgentFactory.Run(ctx, subAgent, task)
	if err != nil {
		tc.LogErr("SubAgent: sub-agent failed", "error", err)
		return domain.ToolResult{Output: fmt.Sprintf("Sub-agent failed: %s", err.Error())}, nil
	}

	// Extract final response from sub-agent turns
	subTurns := tc.AgentFactory.GetTurns(subAgent)
	finalText := extractFinalResponse(subTurns)

	if finalText != "" {
		tc.Log("SubAgent: completed with response", "response_length", len(finalText), "turns", len(subTurns))
		return domain.ToolResult{Output: finalText}, nil
	}

	// Fallback: count tool calls
	toolCalls := 0
	for _, turn := range subTurns {
		if turn.Role == "assistant" {
			if blocks, ok := turn.Content.([]domain.ContentBlock); ok {
				for _, b := range blocks {
					if b.Type == "tool_use" {
						toolCalls++
					}
				}
			}
		}
	}
	if toolCalls == 0 {
		tc.Log("SubAgent: completed (no tool calls)", "turns", len(subTurns))
		return domain.ToolResult{Output: "Task completed"}, nil
	}
	tc.Log("SubAgent: completed with tool calls", "tool_calls", toolCalls, "turns", len(subTurns))
	return domain.ToolResult{Output: fmt.Sprintf("%d operations", toolCalls)}, nil
}

func extractFinalResponse(turns []domain.MessageParam) string {
	for i := len(turns) - 1; i >= 0; i-- {
		if turns[i].Role == "assistant" {
			if blocks, ok := turns[i].Content.([]domain.ContentBlock); ok {
				for _, b := range blocks {
					if b.Type == "text" && strings.TrimSpace(b.Text) != "" {
						return strings.TrimSpace(b.Text)
					}
				}
			}
		}
	}
	return ""
}
